### ISSUE TITLE
Improve pandoc lua filters handling to make it more usable from other package

### ISSUE THREAD

**Issue by cderv** [rstudio/rmarkdown#1897]
We are currently adding some lua filter also in https://github.com/rstudio/bookdown/pull/940 

There are several pieces missing to make things easier to extend what is done in **rmarkdown**

1. `pandoc_lua_filters` helper is currently only working for **rmarkdown** filter. We currently create the same helper for each 📦 

2. The lua filters are not including the same way in all 📦 so it is not easy to make another package filter run before or after another one. 

About 2, sometime it is add inside the `pre_processor` part of the function https://github.com/rstudio/rmarkdown/blob/284b685c2feb68083b3434c6ca38717cd61cbeea/R/html_document_base.R#L109-L110

and in others outside of the pre preprocessor in the main format function, **before** `pandoc_args` argument
https://github.com/rstudio/rmarkdown/blob/284b685c2feb68083b3434c6ca38717cd61cbeea/R/pdf_document.R#L145-L149

We could start fixing this by being more aligned in all format. 

And to improve, I think we need a mechanism for choosing the order of the filter (pandoc will run in the order of the command line). This is necessary to avoir conflicts. There are several ways to that IMO - one is to make lua filters activate with a switch.

**Comment by atusy** [comment]
I think lua filters should be stored in the `pandoc` element of output formats.

```r
fmt <- output_format(pandoc = list(
  lua_filters = c('foo.lua', 'bar.lua'),
  ...
))
```

Then, the `render` function converts `fmt$pandoc$lua_filters` to pandoc arguments.
In this way, downstream formats can easily add, remove, and reorder lua filters.

**Comment by cderv** [comment]
@atusy  need your though on this

**TL;DR: Should we prepend or append lua filters between a output format and its base ?**


By default, `pandoc` argument in `output_format` will be merged with the `base_format` provided, in a way that the overwrite with overlay value, even default. That means new lua filter provided will overwrite the ones from the base format, unless we decide otherwise.

Dummy example to explain (don't run)
```r
# A format using a filter
foo_document<- function(...) {
  output_format(
    pandoc = pandoc_options(lua_filters = "foo.lua")
  )
}

# new format based on this other
bar_document <- function(...) {
  output_format(
    pandoc = pandoc_options(
      lua_filters = "bar.lua"
    ),
    base_format = foo_document()
  )
}
```
Only `bar.lua` would be used because of how a format and its base format are merged. We don't want that and want to keep the one from base too. 

This is the default behavior, and in fact only `args` options in pandoc is not overwritten but appended. In my current branch, I think I'll add a special treatment for `lua_filters` too because we at least want to keep the one of the base format (and not remove then). 

However, questions: 

* By default, do you think lua filters from the base format should be run before or after the one from the format ? It means should we append or prepend ? 
```bash
--lua-filter foo.lua --lua-filter bar.lua
```
or 
```bash
--lua-filter bar.lua --lua-filter foo.lua
``` 

To change the default, one would have to use this to make it works (if default is append)
```r
# A format using a filter
foo_document<- function(...) {
  output_format(
    pandoc = pandoc_options(lua_filters = "foo.lua")
  )
}

# new format based on this other
bar_document <- function(...) {
  from_foo <- foo_document()
  # prepend
  from_foo$pandoc$lua_filters <- c("bar.lua", from_foo$pandoc$lua_filters)
  from_foo
}
```

**Comment by atusy** [comment]
I would chose the "append" strategy if we do not have special reasons to prepend because lua filters act as if functions and "append" is the default merging strategy for functions:

- In most case, `merge_function_outputs()` applies the base before the overlay (`pre_knit`, `post_knit`, `pre_processor`, and `intermediates_generator`).
- Exception is `merge_post_processors()`.

The exception have been introduced intentionally by @jmcphers .
I do not know why and do find no related issues.
https://github.com/rstudio/rmarkdown/commit/cb2537ae286cb9128f1a931183bd60cce5099867

Maybe we should ask him or @yihui for historical background.

**Comment by atusy** [comment]
In the PR #1879, I should have added

```r
args <- c(args, pandoc_lua_filters("number-sections.lua"))
```

before 

```r
args <- c(args, pandoc_args)
```

In terms of consistency with other addition of `args` (see below).
This means user-specified lua filters come after document-builtin lua filters.
It also implies that the "append" strategy provides more consistency for users and developers.

https://github.com/rstudio/rmarkdown/blob/284b685c2feb68083b3434c6ca38717cd61cbeea/R/md_document.R#L51-L64

**Comment by cderv** [comment]
> This means user-specified lua filters come after document-builtin lua filters.

By "user specified" you mean one provided in `pandoc_args` in the output format function? 🤔 
Maybe I should set the filter earlier in `render`. 

You can comment on the PR - I appreciate your great feedback !

**Comment by atusy** [comment]
> By "user specified" you mean one provided in pandoc_args in the output format function? thinking

yes.

> You can comment on the PR

Right. This time, I added comments here so that context is obvious for anybody.
I also referenced the comment from my review of your PR.
Thanks for accepting it.
